react-custom-hooks/client/src/components/useTodos.js

```diff
@@ -0,0 +1,119 @@
+/* eslint-disable no-unused-vars -- Remove me */
+import { useEffect, useState } from 'react';
+
+/**
+ * Manages the Todos by reading from and writing to the backend API.
+ *
+ * @returns {Object} with the following properties:
+ *   isLoading: a boolean indicating if the data has not yet loaded;
+ *   error: any error that occurred
+ *   todos: a list of To Do items;
+ *   addTodo: a function (newTodo) that will add a new To Do;
+ *   toggleCompleted: a function (todoId) that will toggle the completion of an existing To Do.
+ */
+export function useTodos() {
+  const [todos, setTodos] = useState();
+  const [error, setError] = useState();
+  useEffect(() => {
+    /* TODO: If `todos` hasn't been defined yet, read the items from the API
+     * and set them into the `todos` state.
+     */
+    async function getTodos() {
+      try {
+        const todos = await readTodos();
+        setTodos(todos);
+      } catch (err) {
+        setError(err);
+      }
+    }
+    if (!todos) {
+      getTodos();
+    }
+  }, [todos]);
+
+  async function addTodo(newTodo) {
+    /* TODO: Call the API function that creates a Todo item.
+     * When the promise returned by that function resolves, update the `todos` state.
+     * Note that it is critical that you pass a _new_ array. Do not modify the `todos` array.
+     */
+    try {
+      const newItem = await createTodo(newTodo);
+      setTodos([...todos, newItem]);
+    } catch (err) {
+      setError(err);
+    }
+  }
+
+  async function toggleCompleted(todoId) {
+    /* TODO: Find the Todo item being updated, toggle its completed prop, and call
+     * the API function that updates a Todo item.
+     * When the promise returned by that function resolves, update the `todos` state.
+     * When updating this state, use the updated `todo` returned from the API.
+     * Note that it is critical that you pass a _new_ array. Do not modify the `todos` array.
+     */
+    const item = todos.find((element) => element.todoId === todoId);
+    if (item) {
+      item.isCompleted = !item.isCompleted;
+    } else {
+      return;
```

Comment:
This doesn't let the caller know that something was wrong with the request. You should throw or set an error.